eclair-core/src/main/scala/fr/acinq/eclair/crypto/keymanager/LocalOnchainKeyManager.scala

```diff
@@ -0,0 +1,165 @@
+  // - to generate a watch-only wallet with any BIP84-compatible bitcoin wallet
+  // - to generate descriptors that can be import into Bitcoin Core to create a watch-only wallet which can be used
+  // by Eclair to fund transactions (only Eclair will be able to sign wallet inputs).
+
+  // m / purpose' / coin_type' / account' / change / address_index
+  private val master = DeterministicWallet.generate(seed)
+  private val fingerprint = DeterministicWallet.fingerprint(master) & 0xFFFFFFFFL
+  private val fingerPrintHex = String.format("%8s", fingerprint.toHexString).replace(' ', '0')
+  // root bip32 onchain path
+  // we use BIP84 (p2wpkh) path: 84'/{0'/1'}
+  private val rootPath = chainHash match {
+    case Block.RegtestGenesisBlock.hash | Block.TestnetGenesisBlock.hash | Block.SignetGenesisBlock.hash => "84'/1'"
+    case Block.LivenetGenesisBlock.hash => "84'/0'"
+    case _ => throw new IllegalArgumentException(s"invalid chain hash ${chainHash}")
+  }
+  private val rootKey = DeterministicWallet.derivePrivateKey(master, KeyPath(rootPath))
+
+
+  override def getOnchainMasterPubKey(account: Long): String = {
+    val prefix = chainHash match {
+      case Block.RegtestGenesisBlock.hash | Block.TestnetGenesisBlock.hash | Block.SignetGenesisBlock.hash => vpub
+      case Block.LivenetGenesisBlock.hash => zpub
+      case _ => throw new IllegalArgumentException(s"invalid chain hash ${chainHash}")
+    }
+    // master pubkey for account 0 is m/84'/{0'/1'}/0'
+    val accountPub = DeterministicWallet.publicKey(DeterministicWallet.derivePrivateKey(rootKey, hardened(account)))
+    DeterministicWallet.encode(accountPub, prefix)
+  }
+
+  override def getDescriptors(account: Long): Descriptors = {
+    val keyPath = s"$rootPath/$account'".replace('\'', 'h') // Bitcoin Core understands both ' and h suffix for hardened derivation, and h is much easier to parse for external tools
+    val prefix: Int = chainHash match {
+      case Block.LivenetGenesisBlock.hash => xpub
+      case _ => tpub
+    }
+    val accountPub = DeterministicWallet.publicKey(DeterministicWallet.derivePrivateKey(rootKey, hardened(account)))
+    // descriptors for account 0 are:
+    // 84'/{0'/1'}/0'/0/* for main addresses
+    // 84'/{0'/1'}/0'/1/* for change addresses
+    val receiveDesc = s"wpkh([${this.fingerPrintHex}/$keyPath]${encode(accountPub, prefix)}/0/*)"
+    val changeDesc = s"wpkh([${this.fingerPrintHex}/$keyPath]${encode(accountPub, prefix)}/1/*)"
+
+    Descriptors(wallet_name = wallet, descriptors = List(
+      Descriptor(desc = s"$receiveDesc#${descriptorChecksum(receiveDesc)}", internal = false, active = true, timestamp = timestamp.toLong),
+      Descriptor(desc = s"$changeDesc#${descriptorChecksum(changeDesc)}", internal = true, active = true, timestamp = timestamp.toLong),
+    ))
+  }
+
+  override def signPsbt(psbt: Psbt, ourInputs: Seq[Int], ourOutputs: Seq[Int]): Psbt = {
```

Comment:
This should be explicit and return a `Try[Psbt]`: there are a lot of ways this can fail, so the caller really needs to know about it.

I understand why you chose to rely on exceptions (because this is only called inside a future that we want to fail if `signPsbt` didn't succeed), but it makes the code of some of the helper functions below hard to reason about, some of them can be simplified by refactoring to avoid exceptions (I'll give this a try in my own branch).

Reply:
done in https://github.com/ACINQ/eclair/pull/2613/commits/a53c5ec8098e9f80da83ba72278fb2d67783b148